src/Workspaces/Remote/Core/Services/AssetSource.cs

```diff
@@ -13,21 +13,14 @@ namespace Microsoft.CodeAnalysis.Remote
     internal abstract class AssetSource
     {
         private readonly AssetStorage _assetStorage;
-        private readonly int _scopeId;
 
-        protected AssetSource(AssetStorage assetStorage, int scopeId)
+        protected AssetSource(AssetStorage assetStorage)
         {
             _assetStorage = assetStorage;
-            _scopeId = scopeId;
 
-            _assetStorage.RegisterAssetSource(_scopeId, this);
+            _assetStorage.SetAssetSource(this);
```

Comment:
not sure I understand. but asset source will be created by VS's request before it lets others in VS to use OOP. and there is no one else which can create asset source and I dont believe there is a race.